src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataDeserializerProviderProxy.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.Contracts;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.AspNet.OData.Formatter.Deserialization
+{
+    /// <summary>
+    /// The default <see cref="ODataDeserializerProviderProxy"/>.
+    /// </summary>
+    internal partial class ODataDeserializerProviderProxy : ODataDeserializerProvider
+    {
+        private static readonly ODataDeserializerProviderProxy _instance = new ODataDeserializerProviderProxy();
+
+        private IServiceProvider _requestContainer;
+
+        /// <summary>
+        /// Gets the default instance of the <see cref="ODataDeserializerProviderProxy"/>.
+        /// </summary>
+        public static ODataDeserializerProviderProxy Instance
+        {
+            get
+            {
+                return _instance;
+            }
+        }
+
+        public IServiceProvider RequestContainer
+        {
+            get { return _requestContainer; }
+            set
```

Comment:
If it's only supporting to set RequestContainer once, it's better to use the constructer

Reply:
This is existing code refactored into Shared so it's look new but it's now. That does not answer the question but it is a reason to leave it as-is.

Reply:
Also, Its a public API and we're trying not to break back-compat.